### ISSUE TITLE
"Error: container exited immediately"

### ISSUE THREAD

**Issue by MrBenGriffin** [kreuzwerker/terraform-provider-docker#362]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and docker Provider) Version
Terraform v1.1.7 on darwin_amd64 (downloaded binary today).
Docker version 20.10.13, build a224086

### Affected Resource(s)
* `docker_container `

### Terraform Configuration Files
```hcl
terraform {
  required_providers {
    docker = {
      source  = "kreuzwerker/docker"
      version = "~> 2.16.0"
    }
  }
}

provider "docker" {}

# web_nfs is a local image generated via Packer.
# It is definitely recognised during plan.
variable "image_nfs" {
  description = "Image for nfs"
  default     = "ubuntu:web_nfs"
}

resource "docker_image" "nfs" {
  name = var.image_nfs
}

resource "docker_container" "nfs" {
  name  = "nfs"
  image = var.image_nfs
}
```

### Debug Output

(https://gist.github.com/MrBenGriffin/d8234d2478758055b88aeafab91b7f68)

### Expected Behaviour

It would be nice to see an active container that doesn't immediately quit.

### Actual Behaviour
"Error: container exited immediately"

### Steps to Reproduce
Use Mac OS; download the code.
Create a docker image (probably vanilla ubuntu focal will be fine - but I haven't logged in, and that might be a thing).

1. `terraform apply -auto-approve`

### Important Factoids

Nothing remarkable.

**Comment by suzuki-shunsuke** [comment]
I guess the cause of the issue is that the process exits immediately.
Please try to fix container's `command`.

e.g.

```tf
resource "docker_container" "nfs" {
  name  = "nfs"
  image = var.image_nfs
  command = ["tail", "-f", "/dev/null"]
}
```

BTW, please avoid to use private image in the bug report because others can't reproduce the issue with the same code.
https://github.com/kreuzwerker/terraform-provider-docker/blob/master/CONTRIBUTING.md#guide-of-bug-report

**Comment by suzuki-shunsuke** [comment]
https://registry.terraform.io/providers/kreuzwerker/docker/latest/docs/resources/container

Or you may use `tty`. 

e.g.

```tf
resource "docker_container" "ubuntu" {
  name  = "foo"
  image = docker_image.ubuntu.latest
  tty   = true
}

resource "docker_image" "ubuntu" {
  name = "ubuntu:precise"
}
```

**Comment by suzuki-shunsuke** [comment]
I've fixed the example usage of docker_container.
https://github.com/kreuzwerker/terraform-provider-docker/pull/363

**Comment by MrBenGriffin** [comment]
while `tty=true`  works, the documents suggest that this does something else: "_tty (Boolean) If true, allocate a pseudo-tty (docker run -t). Defaults to false._" - this doesn't seem to be relevant 

`must_run = true` -  Does NOT work as documented, and is set to be default true 
"**_must_run** (Boolean) If true, then the Docker container will be kept running. If false, then as long as the container exists, Terraform assumes it is successful. **Defaults to true.**_"

`start=true` -  Does NOT work as documented, and is set to be default true 

"**_start** (Boolean) If true, then the Docker container will be started after creation. If false, then the container is only created. **Defaults to true**._"

Note documents at https://registry.terraform.io/providers/kreuzwerker/docker/latest/docs

https://registry.terraform.io/providers/kreuzwerker/docker/latest/docs/resources/container

**Comment by suzuki-shunsuke** [comment]
To keep the command `bash` running, you have to specify `tty`. Otherwise the process (container) would exit soon.
This is not a problem of this provider.
You can confirm the behavior with Docker CLI.

```console
$ docker run --name foo ubuntu:precise

$ docker ps                           
CONTAINER ID   IMAGE     COMMAND   CREATED   STATUS    PORTS     NAMES

$ docker ps -a | grep foo             
6a62ee268cce   ubuntu:precise                                            "/bin/bash"              9 seconds ago   Exited (0) 8 seconds ago              foo
```

```console
$ docker run --name yoo -t -d ubuntu:precise
579fe376390f0f3dbf1b1a726a39515820049aaf88ebc8f3a0a40a62f3afe1e4

$ docker ps
CONTAINER ID   IMAGE            COMMAND       CREATED         STATUS         PORTS     NAMES
579fe376390f   ubuntu:precise   "/bin/bash"   4 seconds ago   Up 4 seconds             yoo
```

**Comment by MrBenGriffin** [comment]
Yes - what I expect to see is that “must_run" would be the same as the -d -i switches.

Like this:
docker run -d -i --name foo ubuntu:precise

**Comment by suzuki-shunsuke** [comment]
> must_run (Boolean) If true, then the Docker container will be kept running.

This description is wrong. We have to fix this.
If `must_run` is true and the status of the container isn't running, terraform would fail.

**Comment by MrBenGriffin** [comment]
I may be missing something - but I would have thought that must_run would be
docker run -d -i --name foo ubuntu:precise
Whereas tty would be
docker run -d -t --name bar ubuntu:precise

**Comment by suzuki-shunsuke** [comment]
No. `must_run` doesn't change options to run container.
The option change the treat when the container isn't running.

**Comment by MrBenGriffin** [comment]
Okay I think I get this.
Am I right that (when true)

tty =>        docker run  -t ...
stdin_open => docker run  -i ...

So if I want...
docker run  -it ...
then both tty and stdin_open should be set to true, (and they are normally not)?

It seems odd to me, since the idea is to sustain and run an infrastructure, that these options are false by default, since these flags are the normal way to keep a docker container running.

**Comment by MrBenGriffin** [comment]
Also, if must_run is true by default then it follows as a consequence that tty should be true by default.

**Comment by suzuki-shunsuke** [comment]
To keep the process other than shell running, the option `-i` and `-t` isn't needed.
For example, nginx keeps running without `tty`.
So I think if you want to specify `-i` and `-t`, you should specify them explicitly.

**Comment by MrBenGriffin** [comment]
> To keep the process other than shell running, the option -i and -t isn't needed.
> 

I might be ignorant here, but that is not strictly true.

For instance, my nfs image is running an essential nfs daemon service as a part of my infrastructure but it is being ignored by docker.
This may well be my ignorance of docker - as I do not know how to tell docker that it cannot shut down the container unless I use the -it flag.

Likewise I believe this is also the case with a mysql server.
I can see many types of container that I may want as a part of an infrastructure that do require something like -it to stay open.
From where I stand, it would seem that nginx is an exception - not the rule.

But, and in fairness, this may all be due to my own ignorance. 
I do not yet know how to tell docker what is required in order to know that the container is finished with.

----

It would be really helpful if there were documentation with real-world examples - such as how to set up a subnet with various related containers using the docker provider (the current documentation feels very "bare bones”). Maybe something like a load-balanced nginx/apache cluster with a shared r/o nfs mount for the nginx content, using, e.g.
	- a load balancer
	- a collection of nginx/apache servers
	- a shared (r/o) nginx/apache configuration nfs 
	- a shared (r/o) content storage - eg via nfs (web pages used by nginx)
	- a shared home storage
	- a mysql server (or server cluster)

Let me try to explain why.  
Like many others, we are hoping to use terraform to re-engineer, and consider various options for deployment of our infrastructures.
It isn’t just about being able to define an infrastructure - but also to be able to radically change that infrastructure.
Docker is very useful for us, because it allows us to work on these radical changes in a closed, local environment.
This is why we are hoping to use the terraform docker provider, along with other tools such as Packer.

Hashicorp's Terraform marketing is quite persuasive - but there seems to be a cognitive chasm between 
(1) the docs: “Let’s launch an nginx system 
And 
(2) the need: “Let’s define and manage an entire infrastructure of 20 interconnected systems”.

**Comment by suzuki-shunsuke** [comment]
> Likewise I believe this is also the case with a mysql server.

I can keep mysql running without `-i` and `-t` options.

```console
$ docker ps
CONTAINER ID   IMAGE     COMMAND   CREATED   STATUS    PORTS     NAMES

$ docker run --name mysql -e MYSQL_ROOT_PASSWORD=password -d mysql:8.0.28
9deeff9cfff6c69e428998fcfe1593bde45af0c795f5b28a9fa64b2feb405bab

$ docker ps                                                              
CONTAINER ID   IMAGE          COMMAND                  CREATED         STATUS         PORTS                 NAMES
9deeff9cfff6   mysql:8.0.28   "docker-entrypoint.s…"   4 seconds ago   Up 2 seconds   3306/tcp, 33060/tcp   mysql
```

**Comment by MrBenGriffin** [comment]
Excellent,

So - if we are looking at the docker provider, we would implement that with the ‘run’, right?
In fact we need either ‘run’ or ’tty’ - and without those, the container will automatically stop.

Therefore, given ‘must_run’ is true by default, would it not be correct for the docker provider to kick out a warning or error if neither are set?

**Comment by suzuki-shunsuke** [comment]
What do you mean `run`?
It means the option `command`, right? https://registry.terraform.io/providers/kreuzwerker/docker/latest/docs/resources/container#optional

**Comment by suzuki-shunsuke** [comment]
You can keep mysql container running without `tty` and `command`.

```tf
resource "docker_image" "mysql" {
  name = "mysql:8.0.28"
}

resource "docker_container" "mysql" {
  name  = "mysql"
  image = docker_image.mysql.latest
  env   = ["MYSQL_ROOT_PASSWORD=password"]
}
```

**Comment by MrBenGriffin** [comment]
Okay I see that I am confused - and I am grateful to you for your time.

However,  and I am attempting to use terraform via packer, 
I see (cf. https://www.packer.io/plugins/builders/docker#run_command ) 

_An array of arguments to pass to docker run in order to run the container. **By default this is set** to ["-d", "-i", "-t", "--entrypoint=/bin/sh", "--", "{{.Image}}"]_

This is obviously where the wheels are coming off of my bike.
**The Packer defaults are being ignored by Terraform.**
How do I expose the Packer run_command to Terraform?

**Comment by suzuki-shunsuke** [comment]
I'm sorry but I can't understand your situation.

https://www.packer.io/plugins/builders/docker#run_command
This document is for building Docker image with Packer.
In this document, Terraform isn't used.
Could you tell me the example of Package Configuration?

**Comment by MrBenGriffin** [comment]
> I'm sorry but I can't understand your situation.

I will try to explain

TL;DR:- how to ensure that Packer directives are being ingested by Terraform.

If you have a look at https://learn.hashicorp.com/tutorials/terraform/packer
This suggests that we can use Terraform, and provision the infrastructure via Packer.
The tutorial is for AWS - but surely this should also work just as well for Docker, right?

Packer's Docker Builder talks about configuring run_command to the image (which is the basis of containers).
Passing the entrypoint option this way will make it the default entrypoint of the resulting image, […] 
If your docker image embeds a binary intended to be run often, you should consider changing the default entrypoint to point to it.
As I said previously, run_command has a default value - ["-d", "-i", "-t", "--entrypoint=/bin/sh", "--", "{{.Image}}”]

Given that HashiCorp suggests I can provision Terraform via Packer, I would imagine that somehow the run_command from Packer can be ingested by Terraform.
Yet this certainly isn’t happening behind the scenes.

I guess I can do this by exporting the run_command into Terraform, by using Packer’s "local shell post processor" to rewrite (eg via sed) using a “command" in the docker_container.
It seems mysterious though, that I would need to use a scripting language to manage the Terraform scripts.

It would probably be better to generate some form of variables file which can then be ingested by my Terraform script.

**Comment by suzuki-shunsuke** [comment]
https://learn.hashicorp.com/tutorials/terraform/packer

In this tutorial, Packer and Terraform are being used as the following.

1. Build AMI with Packer
2. Edit Terraform Configuration **manually**
3. Create EC2 instance using AMI built with Packer

So Packer and Terraform aren't integrated directly.

---

You can build Docker image with Packer and create container with the image using Terraform.

---

I think you misunderstand Packer's [run_command](https://www.packer.io/plugins/builders/docker#run_command).
This isn't built image's CMD and ENTRYPOINT.
To change built image's CMD and ENTRYPOINT, I guess you have to use [changes](https://www.packer.io/plugins/builders/docker#changes).

**Comment by MrBenGriffin** [comment]
Again, thank-you so much for your time and support.
I accept what you say - and apologise if my confusion and ignorance caused any tension or stress.
Your guidance has been very useful.

**Comment by Junkern** [comment]
Closing because it is resolved
